src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,214 @@
+| (n+1) := conj_transpose_circulant
+
+lemma map_circulant [has_sub n] (v : n → α) (f : α → β) :
+  (circulant v).map f = circulant (λ i, f (v i)) :=
+ext $ λ _ _, rfl
+
+lemma circulant_neg [has_neg α] [has_sub n] (v : n → α) :
+  circulant (- v) = - circulant v :=
+ext $ λ _ _, rfl
+
+@[simp] lemma circulant_zero (α n) [has_zero α] [has_sub n] :
+  circulant 0 = (0 : matrix n n α) :=
+ext $ λ _ _, rfl
+
+lemma circulant_add [has_add α] [has_sub n] (v w : n → α) :
+  circulant (v + w) = circulant v + circulant w :=
+ext $ λ _ _, rfl
+
+lemma circulant_sub [has_sub α] [has_sub n] (v w : n → α) :
+  circulant (v - w) = circulant v - circulant w :=
+ext $ λ _ _, rfl
+
+/-- The product of two circulant matrices `circulant v` and `circulant w` is
+    the circulant matrix generated by `mul_vec (circulant v) w`. -/
+lemma circulant_mul [semiring α] [fintype n] [add_group n] (v w : n → α) :
+  circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w) :=
+begin
+  ext i j,
+  simp only [mul_apply, mul_vec, circulant, dot_product],
+  refine fintype.sum_equiv (equiv.sub_right j) _ _ _,
+  intro x,
+  simp only [equiv.sub_right_apply, sub_sub_sub_cancel_right],
+end
+
+lemma fin.circulant_mul [semiring α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w)
+| 0     := dec_trivial
+| (n+1) := circulant_mul
+
+/-- Multiplication of circulant matrices commutes when the elements do. -/
+lemma circulant_mul_comm
+  [comm_semigroup α] [add_comm_monoid α] [fintype n] [add_comm_group n] (v w : n → α) :
+  circulant v ⬝ circulant w = circulant w ⬝ circulant v :=
+begin
+  ext i j,
+  simp only [mul_apply, circulant, mul_comm],
+  refine fintype.sum_equiv ((equiv.sub_left i).trans (equiv.add_right j)) _ _ _,
+  intro x,
+  congr' 2,
+  { simp },
+  { simp only [equiv.coe_add_right, function.comp_app,
+               equiv.coe_trans, equiv.sub_left_apply],
+    abel }
+end
+
+lemma fin.circulant_mul_comm [comm_semigroup α] [add_comm_monoid α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant w ⬝ circulant v
+| 0     := dec_trivial
+| (n+1) := circulant_mul_comm
+
+/-- `k • circulant v` is another circulantcluant matrix `circulant (k • v)`. -/
+lemma circulant_smul [has_sub n] [has_scalar R α] (k : R) (v : n → α) :
+  circulant (k • v) = k • circulant v :=
+by ext; simp
+
+@[simp] lemma circulant_single_one
+  (α n) [has_zero α] [has_one α] [decidable_eq n] [add_group n] :
+  circulant (pi.single 0 1 : n → α) = (1 : matrix n n α) :=
+by { ext i j, simp [one_apply, pi.single_apply, sub_eq_zero] }
+
+@[simp] lemma circulant_single
+  (n) [semiring α] [decidable_eq n] [add_group n] [fintype n] (a : α) :
+  circulant (pi.single 0 a : n → α) = scalar n a :=
+begin
+  ext i j,
+  simp [pi.single_apply, one_apply, sub_eq_zero],
+end
+
+/-- Note we use `↑i = 0` instead of `i = 0` as `fin 0` has no `0`.
+    This means that we cannot state this with `pi.single`
+    as we did with `matrix.circulant_single`. -/
```

Comment:
Our line-length guideline is 100 columns - there's no need to wrap beyond that, especially within the middle of a sentence:
```
/-- Note we use `↑i = 0` instead of `i = 0` as `fin 0` has no `0`.
This means that we cannot state this with `pi.single` as we did with `matrix.circulant_single`. -/
```